File: 0408-valid-word-abbreviation/0408-valid-word-abbreviation.cpp
class Solution {
public:
    bool validWordAbbreviation(string word, string abbr) {
        int i = 0, j = 0;
        int n = word.length(), m = abbr.length();
        
        while(i < n)
        {
            if(word[i] == abbr[j])
            {
                i++;
                j++;
            }
            
            else if(abbr[j] >= '0' && abbr[j] <= '9')
                
            {
                int t = j + 1;
                while(t < m && abbr[t] >= '0' && abbr[t] <= '9')
                {
                    t++;
                }
                
                if(t != j && abbr[j] == '0') return false;
                
                int num = stoi(abbr.substr(j, t - j));
                i += num;
                j = t;
            }
            else
            {
                return false;
            }
        }
        
        return i == n && j == m;
    }
};

File: 0408-valid-word-abbreviation/0408-valid-word-abbreviation.py
class Solution:
    def validWordAbbreviation(self, word: str, abbr: str) -> bool:
        m, n = len(word), len(abbr)
        
        i, j = 0, 0
        
        while i < m and j < n:
            if abbr[j] >= '0' and abbr[j] <= '9':
                
                if abbr[j] == '0':
                    return False
                
                else:
                    curt = 0
                    while j < n and abbr[j] >= '0' and abbr[j] <= '9':
                        curt = curt * 10 + ord(abbr[j]) - ord('0')
                        j += 1
                    
                    i += curt
            else:
                if word[i] == abbr[j]:
                    i += 1
                    j += 1
                else:
                    return False
        
        return i == m and j == n
                